Creating the correct grid of nodes parallel to the cube edges

tanya-kta opened the pull request:
The grid parallel to the cube edges is being created correctly now, due to direction vectors are perpendicular to the edges. Other grids still have problems with nodes creation.

tanya-kta committed:
Allocate `SimulationMap::nodes` memory before node grid creation; minor other bugs fixed

tanya-kta committed:
Defined cube's faces according to `Face::Face` docstring

tanya-kta committed:
Added settings of the grid parallel to the cube edges

kolayne commented on src/Polyhedron.cu:
```diff
@@ -78,6 +78,25 @@ __host__ __device__ int Polyhedron::get_n_of_faces() const
+            SpacePoint a = faces[i].get_vertices()[j] - faces[i].get_vertices()[0];
+            SpacePoint b = faces[i].get_vertices()[j + 1] - faces[i].get_vertices()[0];
+
+            double sign_of_square = (b % a) * faces[i].get_normal() / abs((b % a) * faces[i].get_normal());
```
Minor: I would write it as
```suggestion
            double sign_of_square = (b % a) * faces[i].get_normal();
            sign_of_square /= abs(sign_of_square);
```
to avoid the duplication, but I think, the current version is ok too

kolayne replied:
I think you'd better rewrite this as
```suggestion
            SpacePoint a = faces[i].get_vertices()[j + 1] - faces[i].get_vertices()[0];
            SpacePoint b = faces[i].get_vertices()[j] - faces[i].get_vertices()[0];

            double sign_of_square = (a % b) * faces[i].get_normal() / abs((a % b) * faces[i].get_normal());
```
To suppress the waning from CLion about arguments' order for `operator%`

kolayne commented on src/SimulationMap.cu:
```diff
@@ -22,23 +22,36 @@ __device__ double const mapnode_dist = 2 * jc::speed;
 {
+    /*
+     * Maximum number of nodes
+     * It must be greater or equal than real number of nodes (`n_of_nodes`) after creation node grid
```
It's surely not "equal than" :)
```suggestion
     * It must be greater than or equal to the real number of nodes (`n_of_nodes`) after creation node grid
```

kolayne commented on src/SimulationMap.cu:
```diff
@@ -22,23 +22,36 @@ __device__ double const mapnode_dist = 2 * jc::speed;
     n_of_nodes = 1;
-
+    /*
     // Direction vector from first node to its top neighbor sets randomly
+    double dir_angle = M_PI / 3;
     SpacePoint direction_vector = relative_point_rotation(start_node_coordinates, start_face->get_vertices()[0],
-                                                          start_face->get_normal(), M_PI * 2 * rand0to1());
+                                                          start_face->get_normal(), M_PI * 2 * rand0to1() * 0 + dir_angle)
+                                  - start_node_coordinates;
+    */
+
+    // Direction vector parallel to the vertical edges of cube.
```
Should this be here for some reason or did you forget to remove it?

kolayne commented on src/SimulationMap.cu:
```diff
@@ -79,28 +92,27 @@ __device__ SimulationMap::SimulationMap(Polyhedron *polyhedron) :
     // Creating new nodes until it can be done, some nodes may have less neighbors than four
     for(int current_node_id = 0; current_node_id < n_of_nodes; ++current_node_id)
     {
-        MapNode &current_node = nodes[current_node_id];
```
I think this is not a positive change. Please, explain or roll back

kolayne commented on src/SimulationMap.cu:
```diff
@@ -234,10 +247,9 @@ __device__ int SimulationMap::get_neighbor_node_id(int current_node_id, SpacePoi
+        nodes[n_of_nodes].detach_particle();
         nodes[n_of_nodes] = MapNode(polyhedron, next_face, neighbor_coordinates);
 
-        *nodes_directions = device_realloc(*nodes_directions, n_of_nodes, n_of_nodes + 1);
```
Because you don't need to update `*nodes_directions` anymore, you should accept `SpacePoint *nodes_directions` instead of `SpacePoint **nodes_directions` as an argument to this function

kolayne reviewed (changes requested):
Cool, but I've found some issues. Keep going!

tanya-kta committed:
Applied changes suggested during the review

Co-authored-by: Nikolay Nechaev <Nikolay_Nechaev@mail.ru>

tanya-kta committed:
Fixed issues found during the review

Rolled back the deletion of variable `current_node`; changed type of `nodes_directions` used in `SimulationMap` methods

Co-authored-by: Nikolay Nechaev <Nikolay_Nechaev@mail.ru>
